timemachine/md/smc.py

```diff
@@ -19,144 +19,190 @@
 IndexArray = Array
 LogWeights = Array
 First = bool
+Iteration = int
 
 BatchPropagator = Callable[[Samples, Lambda], Samples]
-BatchLogProb = Callable[[Samples, Lambda], LogWeights]
-BatchLogProbASMC = Callable[[Samples, Lambda, First], LogWeights]
+BatchLogProb = Callable[[Samples, Lambda, First], LogWeights]
+FindNextLambda = Callable[[Samples, Lambda, Iteration, LogWeights], tuple[Lambda, LogWeights]]
 
 Resampler = Callable[[LogWeights], tuple[IndexArray, LogWeights]]
 ResultDict = dict[str, Any]
 
 
+class SMCMaxIterError(Exception):
+    """
+    Exception when SMC exceeds the maximum number of iters.
+    """
+
+    pass
+
+
 def sequential_monte_carlo(
     samples: Samples,
-    lambdas: Array,
     propagate: BatchPropagator,
     log_prob: BatchLogProb,
     resample: Resampler,
+    find_next_lambda: FindNextLambda,
+    store_intermediate_traj=True,
 ) -> ResultDict:
-    """barebones implementation of Sequential Monte Carlo (SMC)
+    """Implementation of Adaptive Sequential Monte Carlo (SMC).
+       This will adaptively interpolate between lambda=0 and lambda=1,
+       starting at lambda=0.
 
     Parameters
     ----------
     samples: [N,] list
-    lambdas: [K,] array
     propagate: function
         [move(x, lam) for x in xs]
         for example, move(x, lam) might mean "run 100 steps of all-atom MD targeting exp(-u(., lam)), initialized at x"
     log_prob: function
-        [exp(-u(x, lam)) for x in xs]
+        [exp(-u(x, lam, first: bool)) for x in xs]
+        first is set to True for the start of each SMC iteration.
+        This may be used to improve performance by caching prefactors.
     resample: function
         (optionally) perform resampling given an array of log weights
-
+    store_intermediate_traj:
+        Set to True (default) to store intermediate trajectories.
+        Can be set to False to reduce memory requirements.
     Returns
     -------
     trajs_dict
-        "sample_traj"
-            [K-1, N] list of snapshots
+        "traj"
+            [K-1, N] list of snapshots only if `store_intermediate_traj` = True.
+            [1, N] list of snapshots if `store_intermediate_traj` = False.
         "incremental_log_weights_traj"
             [K-1, N] array of incremental log weights
         "ancestry_traj"
             [K-1, N] array of ancestor idxs
         "log_weights_traj"
             [K, N] array of accumulated log weights
+        "lambdas_traj"
+            [K] array of adaptive lambdas
 
     References
     ----------
-    * Arnaud Doucet's annotated bibliography of SMC
-        https://www.stats.ox.ac.uk/~doucet/smc_resources.html
-    * [Dai, Heng, Jacob, Whiteley, 2020] An invitation to sequential Monte Carlo samplers
-        https://arxiv.org/abs/2007.11936
```

Comment:
We may still want to retain the literature references from the previous docstring

Reply:
Added back